src/Prefetch.cpp

```diff
@@ -0,0 +1,228 @@
+#include <algorithm>
+#include <map>
+#include <string>
+
+#include "Prefetch.h"
+#include "IRMutator.h"
+#include "Bounds.h"
+#include "Scope.h"
+#include "Util.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::map;
+using std::string;
+using std::vector;
+
+namespace {
+
+// We need to be able to make loads from a buffer that refer to the
+// same original image/param/etc. This visitor finds a load to the
+// buffer we want to load from, and generates a similar load, but with
+// different args.
+class MakeSimilarLoad : public IRVisitor {
+public:
+    const std::string &buf_name;
+    const std::vector<Expr> &args;
+    Expr load;
+
+    MakeSimilarLoad(std::string name, const std::vector<Expr> &args)
+        : buf_name(name), args(args) {}
+
+private:
+    using IRVisitor::visit;
+
+    void visit(const Call *op) {
+        if (op->name == buf_name) {
+            load = Call::make(op->type, op->name, args, op->call_type, op->func, op->value_index, op->image, op->param);
+        } else {
+            IRVisitor::visit(op);
+        }
+    }
+};
+
+Expr make_similar_load(Stmt s, const std::string &name, const std::vector<Expr> &args) {
+    MakeSimilarLoad v(name, args);
+    s.accept(&v);
+    return v.load;
+}
+
+// Build a Box representing the bounds of a buffer.
+Box buffer_bounds(const string &buf_name, int dims) {
+    Box bounds;
+    for (int i = 0; i < dims; i++) {
+        string dim_name = std::to_string(i);
+
+        Expr buf_min_i = Variable::make(Int(32), buf_name + ".min." + dim_name);
+        Expr buf_extent_i = Variable::make(Int(32), buf_name + ".extent." + dim_name);
+        Expr buf_max_i = buf_min_i + buf_extent_i - 1;
+
+        bounds.push_back(Interval(buf_min_i, buf_max_i));
+    }
+    return bounds;
+}
+
+class InjectPrefetch : public IRMutator {
+public:
+    InjectPrefetch(const map<string, Function> &e) : env(e) { }
+
+private:
+    const map<string, Function> &env;
+    const vector<Prefetch> *prefetches = nullptr;
+    Scope<Interval> bounds;
+
+private:
+    using IRMutator::visit;
+
+    void visit(const Let *op) {
+        Interval in = bounds_of_expr_in_scope(op->value, bounds);
+        bounds.push(op->name, in);
+        IRMutator::visit(op);
+        bounds.pop(op->name);
+    }
+
+    void visit(const LetStmt *op) {
+        Interval in = bounds_of_expr_in_scope(op->value, bounds);
+        bounds.push(op->name, in);
+        IRMutator::visit(op);
+        bounds.pop(op->name);
+    }
+
+    void visit(const ProducerConsumer *op) {
+        const vector<Prefetch> *old_prefetches = prefetches;
```

Comment:
This is a pretty common pattern in tons of IRMutators in Halide... I'm not opposed to fixing this, but this is far from the only example of this. I think we should leave this to a separate discussion.